Update ServerCapabilities to support work done progress

Some of the  Server Capabilities need to be updated to support work done progress (e.g. `ServerCapabilities.getReferencesProvider()` must return `Either<Boolean, ReferenceOptions>` instead of `Boolean`).

i am not sure, i can create a new case but it seems this is the one i hit when i do this:

https://github.com/eclipse/wildwebdeveloper/blob/master/org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/angular/AngularLanguageServer.java#L52

there i enable experimental ivy stuff, i add the following command arg:

commands.add("--experimental-ivy");

if i do that then i get a lot of exceptions:

feb. 25, 2021 8:37:19 P.M. org.eclipse.lsp4j.jsonrpc.RemoteEndpoint logIssues
WARNING: Issue found in NotificationMessage: Message could not be parsed.
com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Kind of progress notification is empty
	at com.google.gson.Gson.fromJson(Gson.java:947)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.MessageTypeAdapter.fromJson(MessageTypeAdapter.java:329)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.MessageTypeAdapter.parseParams(MessageTypeAdapter.java:249)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.MessageTypeAdapter.read(MessageTypeAdapter.java:119)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.MessageTypeAdapter.read(MessageTypeAdapter.java:55)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	at org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler.parseMessage(MessageJsonHandler.java:119)
	at org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler.parseMessage(MessageJsonHandler.java:114)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:193)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:853)
Caused by: com.google.gson.stream.MalformedJsonException: Kind of progress notification is empty
	at org.eclipse.lsp4j.adapters.WorkDoneProgressNotificationAdapter.createNotification(WorkDoneProgressNotificationAdapter.java:111)
	at org.eclipse.lsp4j.adapters.WorkDoneProgressNotificationAdapter.read(WorkDoneProgressNotificationAdapter.java:97)
	at org.eclipse.lsp4j.adapters.WorkDoneProgressNotificationAdapter.read(WorkDoneProgressNotificationAdapter.java:37)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:199)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	... 15 more

this is because the json message is:

{"jsonrpc":"2.0","method":"$/progress","params":{"token":"ngcc","value":{"done":false,"configFilePath":"c:/Users/jcomp/git/servoy_master/servoy-eclipse/com.servoy.eclipse.ngclient.ui/node/tsconfig.json","message":"Running ngcc for project c:/Users/jcomp/git/servoy_master/servoy-eclipse/com.servoy.eclipse.ngclient.ui/node/tsconfig.json"}}}

that code in 
org.eclipse.lsp4j.adapters.WorkDoneProgressNotificationAdapter.read(WorkDoneProgressNotificationAdapter.java

can't handle that 
the quick fix would be i think that we also in the switch:

switch (name) {

have a case

case "done": {
  if (in.nextBoolean()) kind = "report";
  else kind = "end";
  break;
}

at least i think we get a "done": "false" and a "done": "true" with then a message?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,22 @@
 
 Fixed issues: https://github.com/eclipse/lsp4j/milestone/18?closed=1
 
+Breaking API changes:
+ * Type of `ServerCapabilities.hoverProvider` changed from `Boolean` to `Either<Boolean, HoverOptions>`
+ * Type of `ServerCapabilities.definitionProvider` changed from `Boolean` to `Either<Boolean, DefinitionOptions>`
+ * Type of `ServerCapabilities.typeDefinitionProvider` changed from `Either<Boolean, StaticRegistrationOptions>` to `Either<Boolean, TypeDefinitionRegistrationOptions>`
+ * Type of `ServerCapabilities.implementationProvider` changed from `Either<Boolean, StaticRegistrationOptions>` to `Either<Boolean, ImplementationRegistrationOptions>`
+ * Type of `ServerCapabilities.referencesProvider` changed from `Boolean` to `Either<Boolean, ReferenceOptions>`
+ * Type of `ServerCapabilities.documentHighlightProvider` changed from `Boolean` to `Either<Boolean, DocumentHighlightOptions>`
+ * Type of `ServerCapabilities.documentSymbolProvider` changed from `Boolean` to `Either<Boolean, DocumentSymbolOptions>`
+ * Type of `ServerCapabilities.workspaceSymbolProvider` changed from `Boolean` to `Either<Boolean, WorkspaceSymbolOptions>`
+ * Type of `ServerCapabilities.documentFormattingProvider` changed from `Boolean` to `Either<Boolean, DocumentFormattingOptions>`
+ * Type of `ServerCapabilities.documentRangeFormattingProvider` changed from `Boolean` to `Either<Boolean, DocumentRangeFormattingOptions>`
+ * Type of `ServerCapabilities.declarationProvider` changed from `Either<Boolean, StaticRegistrationOptions>` to `Either<Boolean, DeclarationRegistrationOptions>`
+ * Type of `ServerCapabilities.callHierarchyProvider` changed from `Either<Boolean, StaticRegistrationOptions>` to `Either<Boolean, CallHierarchyRegistrationOptions>`
+ * Type of `ServerCapabilities.selectionRangeProvider` changed from `Either<Boolean, StaticRegistrationOptions>` to `Either<Boolean, SelectionRangeRegistrationOptions>`
+ * Type of `ServerCapabilities.linkedEditingRangeProvider` changed from `Either<Boolean, StaticRegistrationOptions>` to `Either<Boolean, LinkedEditingRangeRegistrationOptions>`
+
 Deprecated API changes:
  * In DAP Protocol `VariablePresentationHintKind.DATA_BREAKPOINT` has been marked @Deprecated to match its deprecation in the protocol in 1.44.
  * In DAP Protocol `SetBreakpointsArguments.lines` has been marked @Deprecated to match its deprecation in protocol many versions ago.

--- a/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend
+++ b/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend
@@ -2239,6 +2239,35 @@ class CodeActionOptions extends AbstractWorkDoneProgressOptions {
 }
 
 /**
+ * Code Action registration options.
+ */
+@JsonRpcData
+class CodeActionRegistrationOptions extends AbstractTextDocumentRegistrationAndWorkDoneProgressOptions {
+	/**
+	 * CodeActionKinds that this server may return.
+	 *
+	 * The list of kinds may be generic, such as `CodeActionKind.Refactor`, or the server
+	 * may list out every specific kind they provide.
+	 */
+	List<String> codeActionKinds
+
+	/**
+	 * The server provides support to resolve additional
+	 * information for a code action.
+	 *
+	 * Since 3.16.0
+	 */
+	Boolean resolveProvider
+
+	new() {
+	}
+
+	new(List<String> codeActionKinds) {
+		this.codeActionKinds = codeActionKinds
+	}
+}
+
+/**
  * Code Lens options.
  */
 @JsonRpcData
@@ -2916,6 +2945,20 @@ class DocumentFormattingParams implements WorkDoneProgressParams {
 }
 
 /**
+ * Document formatting options.
+ */
+@JsonRpcData
+class DocumentFormattingOptions extends AbstractWorkDoneProgressOptions {
+}
+
+/**
+ * Document formatting registration options.
+ */
+@JsonRpcData
+class DocumentFormattingRegistrationOptions extends AbstractTextDocumentRegistrationAndWorkDoneProgressOptions {
+}
+
+/**
  * A document highlight is a range inside a text document which deserves special attention. Usually a document highlight
  * is visualized by changing the background color of its range.
  */
@@ -3026,7 +3069,7 @@ class DocumentLinkParams extends WorkDoneProgressAndPartialResultParams {
  * Document link options
  */
 @JsonRpcData
-class DocumentLinkOptions {
+class DocumentLinkOptions extends AbstractWorkDoneProgressOptions {
 	/**
 	 * Document links have a resolve provider as well.
 	 */
@@ -4853,7 +4896,7 @@ class ServerCapabilities {
 	/**
 	 * The server provides hover support.
 	 */
-	Boolean hoverProvider
+	Either<Boolean, HoverOptions> hoverProvider
 
 	/**
 	 * The server provides completion support.
@@ -4868,41 +4911,41 @@ class ServerCapabilities {
 	/**
 	 * The server provides goto definition support.
 	 */
-	Boolean definitionProvider
+	Either<Boolean, DefinitionOptions> definitionProvider
 
 	/**
 	 * The server provides Goto Type Definition support.
 	 *
 	 * Since 3.6.0
 	 */
-	Either<Boolean, StaticRegistrationOptions> typeDefinitionProvider
+	Either<Boolean, TypeDefinitionRegistrationOptions> typeDefinitionProvider
 
 	/**
 	 * The server provides Goto Implementation support.
 	 *
 	 * Since 3.6.0
 	 */
-	Either<Boolean, StaticRegistrationOptions> implementationProvider
+	Either<Boolean, ImplementationRegistrationOptions> implementationProvider
 
 	/**
 	 * The server provides find references support.
 	 */
-	Boolean referencesProvider
+	Either<Boolean, ReferenceOptions> referencesProvider
 
 	/**
 	 * The server provides document highlight support.
 	 */
-	Boolean documentHighlightProvider
+	Either<Boolean, DocumentHighlightOptions> documentHighlightProvider
 
 	/**
 	 * The server provides document symbol support.
 	 */
-	Boolean documentSymbolProvider
+	Either<Boolean, DocumentSymbolOptions> documentSymbolProvider
 
 	/**
 	 * The server provides workspace symbol support.
 	 */
-	Boolean workspaceSymbolProvider
+	Either<Boolean, WorkspaceSymbolOptions> workspaceSymbolProvider
 
 	/**
 	 * The server provides code actions. The `CodeActionOptions` return type is only
@@ -4919,12 +4962,12 @@ class ServerCapabilities {
 	/**
 	 * The server provides document formatting.
 	 */
-	Boolean documentFormattingProvider
+	Either<Boolean, DocumentFormattingOptions> documentFormattingProvider
 
 	/**
 	 * The server provides document range formatting.
 	 */
-	Boolean documentRangeFormattingProvider
+	Either<Boolean, DocumentRangeFormattingOptions> documentRangeFormattingProvider
 
 	/**
 	 * The server provides document formatting on typing.
@@ -4960,7 +5003,7 @@ class ServerCapabilities {
 	 *
 	 * Since 3.14.0
 	 */
-	Either<Boolean, StaticRegistrationOptions> declarationProvider
+	Either<Boolean, DeclarationRegistrationOptions> declarationProvider
 
 	/**
 	 * The server provides execute command support.
@@ -4989,14 +5032,14 @@ class ServerCapabilities {
 	 *
 	 * Since 3.16.0
 	 */
-	Either<Boolean, StaticRegistrationOptions> callHierarchyProvider
+	Either<Boolean, CallHierarchyRegistrationOptions> callHierarchyProvider
 
 	/**
 	 * The server provides selection range support.
 	 *
 	 * Since 3.15.0
 	 */
-	Either<Boolean, StaticRegistrationOptions> selectionRangeProvider
+	Either<Boolean, SelectionRangeRegistrationOptions> selectionRangeProvider
 
 	/**
 	 * The server provides linked editing range support.
@@ -6559,7 +6602,7 @@ class TextDocumentSaveRegistrationOptions extends TextDocumentRegistrationOption
 }
 
 @JsonRpcData
-class CompletionRegistrationOptions extends TextDocumentRegistrationOptions {
+class CompletionRegistrationOptions extends AbstractTextDocumentRegistrationAndWorkDoneProgressOptions {
 	/**
 	 * Most tools trigger completion request automatically without explicitly requesting
 	 * it using a keyboard shortcut (e.g. Ctrl+Space). Typically they do so when the user
@@ -6617,7 +6660,7 @@ class SignatureHelpRegistrationOptions extends AbstractTextDocumentRegistrationA
 }
 
 @JsonRpcData
-class CodeLensRegistrationOptions extends TextDocumentRegistrationOptions {
+class CodeLensRegistrationOptions extends AbstractTextDocumentRegistrationAndWorkDoneProgressOptions {
 	/**
 	 * Code lens has a resolve provider as well.
 	 */
@@ -6632,7 +6675,7 @@ class CodeLensRegistrationOptions extends TextDocumentRegistrationOptions {
 }
 
 @JsonRpcData
-class DocumentLinkRegistrationOptions extends TextDocumentRegistrationOptions {
+class DocumentLinkRegistrationOptions extends AbstractTextDocumentRegistrationAndWorkDoneProgressOptions {
 	/**
 	 * Document links have a resolve provider as well.
 	 */

--- a/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend
+++ b/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend
@@ -5046,7 +5046,7 @@ class ServerCapabilities {
 	 *
 	 * Since 3.16.0
 	 */
-	Either<Boolean, StaticRegistrationOptions> linkedEditingRangeProvider
+	Either<Boolean, LinkedEditingRangeRegistrationOptions> linkedEditingRangeProvider
 
 	/**
 	 * The server provides semantic tokens support.
@@ -7703,6 +7703,18 @@ class SelectionRangeOptions extends AbstractWorkDoneProgressOptions {
  */
 @JsonRpcData
 class SelectionRangeRegistrationOptions extends AbstractTextDocumentRegistrationAndWorkDoneProgressOptions {
+	/**
+	 * The id used to register the request. The id can be used to deregister
+	 * the request again. See also Registration#id.
+	 */
+	String id
+
+	new() {
+	}
+
+	new(String id) {
+		this.id = id
+	}
 }
 
 /**

